cpp/src/rolling/jit/code/kernel.cpp

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace cudf {
+namespace experimental {
+namespace rolling {
+namespace jit {
+namespace code {
+
+
+
+const char* kernel_headers = 
+R"***(
+#include <cudf/types.hpp>
+)***";
+
+const char* kernel =
+R"***(
+#include "operation.h"
+
+template <typename WindowType>
+cudf::size_type get_window(WindowType window, cudf::size_type index) { return 0; }
+
+template <>
+cudf::size_type get_window(cudf::size_type window, cudf::size_type index) { return window; }
+
+template <>
+cudf::size_type get_window(cudf::size_type* window, cudf::size_type index) {
+  return window[index];
+}
+
+template <typename InType, typename OutType, class agg_op, typename WindowType>
+__global__
+void gpu_rolling_new(cudf::size_type nrows,
+                 InType const* const __restrict__ in_col, 
+                 cudf::bitmask_type const* const __restrict__ in_col_valid,
+                 OutType* __restrict__ out_col, 
+                 cudf::bitmask_type* __restrict__ out_col_valid,
+                 cudf::size_type * __restrict__ output_valid_count,
+                 WindowType preceding_window_begin,
+                 WindowType following_window_begin,
+                 cudf::size_type min_periods)
+{
+  constexpr int warp_size{32};
```

Comment:
That requires stringifying that header. I will look into making it work with Jitify.

Reply:
I looked at doing this, and it requires installing the CUB headers (since cuda.cuh includes CUB headers) with a cuDF install so that we can JIT-compile this header. I could do that, but I don't have time before a month vacation, so I hope you'll agree this can be done later when we add more JIT functionality. OK?

Reply:
OK, no problem, and enjoy your vacation!

Reply:
warp_size is no longer used in the kernel.